api/test/opentelemetry/trace/tracer_test.rb

```diff
@@ -0,0 +1,124 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'test_helper'
+
+describe OpenTelemetry::Trace::Tracer do
+  Propagation = OpenTelemetry::DistributedContext::Propagation
+  Trace = OpenTelemetry::Trace
+  Tracer = Trace::Tracer
+
+  # Tracer to verify expectation that `Span#finish` is called
+  class TestInSpanFinishTracer < Tracer
+    # Override `start_span` to return mock span
+    def start_span(*)
+      Minitest::Mock.new.expect(:finish, nil)
+    end
+  end
+
+  let(:invalid_span) { Trace::Span::INVALID }
+  let(:tracer) { Tracer.new }
+
+  describe '#current_span' do
+    let(:current_span) { tracer.start_span('current') }
+
+    it 'returns an invalid span by default' do
+      tracer.current_span.must_equal(invalid_span)
+    end
+
+    it 'returns the current span' do
+      wrapper_span = tracer.start_span('wrapper')
+
+      tracer.with_span(wrapper_span) do
+        tracer.current_span.must_equal(wrapper_span)
+      end
+    end
+  end
+
+  describe '#in_span' do
+    it 'yields the new span' do
+      tracer.in_span('wrapper') do |span|
+        span.wont_equal(invalid_span)
+        tracer.current_span.must_equal(span)
+      end
+    end
+
+    it 'returns the result of the block' do
+      result = tracer.in_span('wrapper') { 'my-result' }
+      result.must_equal('my-result')
+    end
+
+    it 'finishes the new span at the end of the block' do
+      finish_tracer = TestInSpanFinishTracer.new
+      mock_span = nil
+      finish_tracer.in_span('wrapper') { |span| mock_span = span }
+      mock_span.verify
+    end
+  end
+
+  describe '#with_span' do
+    it 'yields the passed in span' do
+      wrapper_span = tracer.start_span('wrapper')
+
+      tracer.with_span(wrapper_span) do |span|
+        span.must_equal(wrapper_span)
+      end
+    end
+  end
```

Comment:
Both `with_span` and `in_span` should exhibit the following behaviour:
```ruby
span_a = tracer.start_span('a')
span_b = tracer.start_span('b')
tracer.with_span(span_a) do
  tracer.current_span.must_equal(span_a)
  tracer.with_span(span_b) do
    tracer.current_span.must_equal(span_b)
  end
  tracer.current_span.must_equal(span_a)
end
```
i.e. the span that was active before the block is re-activated after the block. We don't have tests for that here.

Reply:
@fbogsany great example! I'll get that in.